File: 백준/Gold/14500. 테트로미노/테트로미노.cc
#include <iostream>
using namespace std;

#define max(a, b) (a > b) ? (a) : (b)

int N, M, ans;
int board[500][500];
short visited[500][500];

int dx[] = { 1, 0, -1, 0 };
int dy[] = { 0, 1, 0, -1 };

void dfs(int x, int y, int move, int res) {
    if (move == 3) {
        ans = max(ans, res);
        return;
    }

    for (int i = 0; i < 4; i++)
    {
        int nx = x + dx[i];
        int ny = y + dy[i];

        if (nx < 0 || nx >= N || ny < 0 || ny >= M || visited[nx][ny] == 1) continue;

        if (move == 1) {
            visited[nx][ny] = 1;
            dfs(x, y, move + 1, res + board[nx][ny]);
            visited[nx][ny] = 0;
        }

        visited[nx][ny] = 1;
        dfs(nx, ny, move + 1, res + board[nx][ny]);
        visited[nx][ny] = 0;
    }
}

int main(void) {
    cin.tie(0);
    cout.tie(0);
    ios_base::sync_with_stdio(false);
    
    cin >> N >> M;

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++) {
            cin >> board[i][j];
        }
    }

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++) {
            visited[i][j] = 1;
            dfs(i, j, 0, board[i][j]);
            visited[i][j] = 0;
        }
    }
    
    cout << ans;

    return 0;
}

File: 백준/Gold/14500. 테트로미노/테트로미노.py
def dfs(num, x, y, sum):
    global visited
    global mx

    if num == 4:
        if mx < sum:
            mx = sum
        return
    
    if mx and mx < sum:
        return

    for move in range(4):
        nx = x + dx[move]
        ny = y + dy[move]

        if 0 <= nx < N and 0 <= ny < M and visited[nx][ny] == False:
            if num == 2:
                visited[nx][ny] = True
                dfs(num + 1, x, y, sum + paper[nx][ny])
                visited[nx][ny] = False

            visited[nx][ny] = True
            dfs(num + 1, nx, ny, sum + paper[nx][ny])
            visited[nx][ny] = False


dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

N, M = map(int, input().split())
paper = [list(map(int, input().split())) for _ in range(N)]
visited = [[False] * M for _ in range(N)]
mx = 0

for i in range(N):
    for j in range(M):
        visited[i][j] = True
        dfs(1, i, j, paper[i][j])
        visited[i][j] = False

print(mx)